lib/loader/filemgr.cpp

```diff
@@ -144,84 +144,96 @@ Expect<uint64_t> FileMgr::readU64() {
   return Result;
 }
 
-// Decode and read a signed int. See "include/loader/filemgr.h".
-Expect<int32_t> FileMgr::readS32() {
+template <typename RetType, size_t N> Expect<RetType> FileMgr::readSN() {
+  static_assert(N >= 8, "The N of S_N must have at least length of a byte");
+  static_assert(8 * sizeof(RetType) >= N,
+                "RetType cannot hold the full range of S_N");
+  static_assert(std::is_signed_v<RetType>,
+                "RetType of S_N must be signed type");
+
   if (Status != ErrCode::Value::Success) {
     return Unexpect(Status);
   }
   // Set the flag to the start offset.
   LastPos = Pos;
 
-  // Read and decode S32.
-  int32_t Result = 0;
-  uint32_t Offset = 0;
-  Byte Byte = 0x80;
-  while (Byte & 0x80) {
-    if (unlikely(Offset >= 32)) {
+  // Read and decode S_N.
+  RetType Result = 0;
+  size_t Offset = 0;
+  size_t RemainingBits = N;
+  while (true) {
+    if (RemainingBits <= 0) {
       Status = ErrCode::Value::IntegerTooLong;
       return Unexpect(Status);
     }
-    if (auto Res = testRead(1); unlikely(!Res)) {
-      return Unexpect(Res);
-    }
-    Byte = Data[Pos++];
-    const int32_t Mask = Offset == 28 ? INT32_C(0x0F) : INT32_C(0x7F);
-    Result |= (Byte & Mask) << Offset;
-    Offset += 7;
-  }
-  if (Offset == 35) {
-    // The signed-extend bits should be the same.
-    if (unlikely(((Byte & 0x70) != 0x70 && (Byte & 0x70) != 0) ||
-                 (Byte & 0x40) >> 6 != (Byte & 0x08) >> 3)) {
-      Status = ErrCode::Value::IntegerTooLarge;
-      return Unexpect(Status);
-    }
-  }
-  if (Byte & 0x40 && Offset < 32) {
-    Result |= static_cast<int32_t>(UINT32_C(0xFFFFFFFF) << Offset);
-  }
-  return Result;
-}
 
-// Decode and read a signed long long int. See "include/loader/filemgr.h".
-Expect<int64_t> FileMgr::readS64() {
-  if (Status != ErrCode::Value::Success) {
-    return Unexpect(Status);
-  }
-  // Set the flag to the start offset.
-  LastPos = Pos;
+    // In the rest logic, RemainingBits must be at least 1.
 
-  // Read and decode S64.
-  int64_t Result = 0;
-  uint64_t Offset = 0;
-  Byte Byte = 0x80;
-  while (Byte & 0x80) {
-    if (unlikely(Offset >= 64)) {
-      Status = ErrCode::Value::IntegerTooLong;
-      return Unexpect(Status);
-    }
-    if (auto Res = testRead(1); unlikely(!Res)) {
+    WasmEdge::Byte Byte;
+    if (auto Res = readByte(); unlikely(!Res)) {
```

Comment:
Please use the `testRead(1)` and `Byte = Data[Pos++]` mechanism as the origin code.
This is for printing the correct error message of the loading offset when error occurs.